README.md
# gen-statsd

[![Go Report Card](https://goreportcard.com/badge/github.com/circonus-labs/gen-statsd)](https://goreportcard.com/report/github.com/circonus-labs/gen-statsd)

gen-statsd is an experimental tool written in Go for load testing with StatsD metrics.

## Configuration

### Command-line Flags

```
  Usage of gen-statsd:
  -agents int
        max number of agents to run concurrently (default 10)
  -counters int
        number of counters for each agent to hold (default 50)
  -flush-interval duration
        how often to flush metrics (default 10s)
  -gauges int
        number of gauges for each agent to hold (default 30)
  -prefix string
        prefix for metrics (default "gen-statsd")
  -protocol string
        network protocol to use, tcp or udp (default "udp")
  -spawn-drift int
        spread new agent generation by 0-n seconds (default 10)
  -statsd-host string
        address of statsD host (default "localhost:8125")
  -tag-format string
        format of the tags to send. accepted values "datadog" or "influx"
  -tags string
        list of K:V comma separated tags. Example: key1:tag1,key2:tag2
  -timers int
        number of timers for each agent to hold (default 20)
  -version
        show version information
```

### Environment Variables

  | Variable      |                       Description                               |
  |:-------------:|:----------------------------------------------------------------|
  |AGENTS         |max number of agents to run concurrently (default 10)            |
  |COUNTERS       |number of counters for each agent to hold (default 50)           |
  |FLUSH_INTERVAL |how often to flush metrics (default 10s)                         |
  |GAUGES         |number of gauges for each agent to hold (default 30)             |
  |PREFIX         |prefix for metrics (default "gen-statsd")                        |
  |PROTOCOL       |network protocol to use, tcp or udp (default "udp")              |
  |SPAWN_DRIFT    |spread new agent generation by 0-n seconds (default 10)          |
  |STATSD_HOST    |address of statsD host (default "localhost:8125")                |
  |TAG_FORMAT     |format of the tags to send. accepted values "datadog" or "influx |
  |TAGS           |list of K:V comma separated tags. Example: key1:tag1,key2:tag2   |
  |TIMERS         |number of timers for each agent to hold (default 20)             |
  |VERSION        |show version information                                         |

## Releases

### Binary

[Releases](https://github.com/circonus-labs/gen-statsd/releases)

### Docker

See [Docker Hub](hub.docker.com/circonuslabs/gen-statsd) for additional details.

* Be sure to have Docker installed on your machine
* Run: 
  * `docker pull circonuslabs/gen-statsd:latest`
  * `docker run -e STATSD_HOST=127.0.0.1:8125 -e AGENTS=1 ... circonuslabs/gen-statsd:latest`
* Note: You can also use the docker flag `--env-file` and provide the path to a file with the environment variables from above defined.

## Installation

### From Source

- Be sure to have `GOBIN` set
- Be sure that the location of GOBIN is in your `PATH`

1. Clone this repo
1. Navigate to the repo in your filesystem
1. Run `go install`

## Contributing

[Contributing Guide](./CONTRIBUTING.md)

CHANGELOG.md
# v1.0.0-alpha.1

* testing release, not guaranteed to be feature complete

agent_test.go
// Copyright © 2020 Circonus, Inc. <support@circonus.com>
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.
//

package main

import (
	"fmt"
	"strconv"
	"testing"
	"time"
)

var agent *Agent

func TestParseTags(t *testing.T) {

	//Test setup
	goodTags := "key1:value1,key2:value2,key3:value3"
	badTags1 := "key1:value1,key2:value2,key3"
	badTags2 := "key1:value1,key2:value2,key3:"

	//Run the tests and verify output
	_, err := parseTags(goodTags)
	if err != nil {
		t.Errorf("Expected no error but got %s", err)
	}

	_, err = parseTags(badTags1)
	if err == nil {
		t.Error("Expected an error but one wasn't returned")
	}

	_, err = parseTags(badTags2)
	if err == nil {
		t.Error("Expected an error but one wasn't returned")
	}
}

func TestGenMetricsNames(t *testing.T) {

	//Test Setup
	id := 0
	metricType := "counter"
	n := 3

	//Run tests and verify output
	metricNames := genMetricsNames(metricType, id, n)
	if len(metricNames) != 3 {
		t.Errorf("Expected length of slice to be %d, got %d", n, len(metricNames))
	}

	for i, name := range metricNames {
		if name != "agent"+strconv.Itoa(id)+"-"+metricType+strconv.Itoa(i) {
			t.Errorf("Expected: %s, got %s", fmt.Sprintf("agent%d-%s%d", id, metricType, i), name)
		}
	}
}

func TestCreateAgent(t *testing.T) {

	//Test Setup
	id := 0
	num := 1
	flush := time.Second * 0
	addr := ":8125"
	prefix := "test"
	goodTags := "key1:value1,key2:value2"
	badTags := "key1:value1,key2:"
	tagFormat1 := "datadog"
	tagFormat2 := "influx"
	badTagFormat := "not a tag format"

	//Run tests and verify output
	_, err := CreateAgent(id, num, num, num, flush, addr, prefix, goodTags, "udp", tagFormat1)
	if err != nil {
		t.Errorf("expected no error, got: %s", err)
	}
	_, err = CreateAgent(id, num, num, num, flush, addr, prefix, badTags, "udp", tagFormat1)
	if err == nil {
		t.Errorf("expected error, got: %s", err)
	}
	_, err = CreateAgent(id, num, num, num, flush, addr, prefix, goodTags, "udp", tagFormat2)
	if err != nil {
		t.Errorf("expected no error, got: %s", err)
	}
	_, err = CreateAgent(id, num, num, num, flush, addr, prefix, goodTags, "udp", badTagFormat)
	if err == nil {
		t.Errorf("expected error, got: %s", err)
	}
}

func TestNewAgentController(t *testing.T) {
	ac := NewAgentController()
	if ac == nil {
		t.Error("expected agent controller got nil")
	}
}

// func TestAgentControllerStart(t *testing.T) {
// 	ac := NewAgentController()
// 	conf := config{
// 		statsdHost:    ":8125",
// 		network:       "udp",
// 		flushInterval: time.Second * 1,
// 		counters:      1,
// 		gauges:        0,
// 		timers:        0,
// 		agents:        1,
// 		spawnDrift:    10,
// 	}
// 	buf := make([]byte, 1024)
// 	go udpListen(buf)
// 	ac.Start(conf)
// 	time.Sleep(time.Second * 2)
// 	ac.ctx.Done()

// 	fmt.Print(string(buf))
// }

// func udpListen(buf []byte) (int, error) {
// 	srv, err := net.ListenPacket("udp", ":8125")
// 	if err != nil {
// 		return 0, err
// 	}
// 	defer srv.Close()
// 	n, _, err := srv.ReadFrom(buf)
// 	if err != nil {
// 		return 0, err
// 	}
// 	return n, nil
// }


agent.go
// Copyright © 2020 Circonus, Inc. <support@circonus.com>
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.
//

package main

import (
	"context"
	"errors"
	"fmt"
	"log"
	"math/rand"
	"os"
	"strings"
	"sync"
	"time"

	statsd "gopkg.in/alexcesaro/statsd.v2"
)

//AgentController is the main controller of the agents
type AgentController struct {
	sig  chan os.Signal
	wg   sync.WaitGroup
	ctx  context.Context
	cncl context.CancelFunc
}

//Agent is a struct for generating and sending StatsD data
type Agent struct {
	id            int
	flushInterval time.Duration
	counterNames  []string
	gaugeNames    []string
	timerNames    []string
	statsdClient  *statsd.Client
}

//NewAgentController creates a new agent pool
func NewAgentController() *AgentController {
	ctx, cancel := context.WithCancel(context.Background())
	return &AgentController{
		sig:  make(chan os.Signal, 1),
		wg:   sync.WaitGroup{},
		ctx:  ctx,
		cncl: cancel,
	}
}

//Start kicks off the main process of sending statsD metrics from agents
func (ac *AgentController) Start(c config) error {

	SignalNotifySetup(ac.sig)
	go HandleSignals(ac.cncl, ac.sig)

	for i := 0; i < c.agents; i++ {
		ac.wg.Add(1)
		go func(id int) {
			agent, err := CreateAgent(id, c.counters, c.gauges, c.timers, c.flushInterval, c.statsdHost, c.prefix, c.tags, c.network, c.tagFormat)
			if err != nil {
				log.Printf("error instantiating agent%d: %s\n", id, err)
				ac.ctx.Done()
				ac.wg.Done()
				return
			}
			log.Printf("launching agent %d\n", id)
			agent.Start(ac.ctx)
			ac.wg.Done()
		}(i)
		if done(ac.ctx) {
			break
		}
		time.Sleep(time.Duration(rand.Intn(c.spawnDrift)) * time.Second)
	}

	ac.wg.Wait()
	return nil
}

//CreateAgent creates a new instance of an Agent
func CreateAgent(id, counters, gauges, timers int, flush time.Duration, addr, prefix, tags, network, tagFormat string) (*Agent, error) {

	//Setup some variables
	var client *statsd.Client
	var tagOption statsd.Option

	//Create the client
	client, err := statsd.New(
		statsd.Address(addr),
		statsd.Network(network),
		statsd.FlushPeriod(flush),
		statsd.Prefix(prefix),
		statsd.ErrorHandler(func(err error) {
			log.Printf("error sending metrics: %s\n", err)
		}),
	)
	if err != nil {
		log.Printf("error creating statsd client: %s\n", err)
	}

	//Check the tagformat
	if tagFormat != "" {
		client, err = parseTagFormat(client, tagFormat)
		if err != nil {
			return nil, err
		}
	}

	//Check for tags
	if tags != "" {
		var err error
		tagOption, err = parseTags(tags)
		if err != nil {
			return nil, err
		}
		client = client.Clone(tagOption)
	}

	a := &Agent{
		id:            id,
		flushInterval: time.Duration(flush),
		counterNames:  genMetricsNames("counter", id, counters),
		gaugeNames:    genMetricsNames("gauge", id, gauges),
		timerNames:    genMetricsNames("timer", id, timers),
		statsdClient:  client,
	}
	return a, nil
}

//Start starts an agent generating and sending metrics to the desired host
func (a *Agent) Start(ctx context.Context) {
	ticker := time.NewTicker(a.flushInterval)
	for {
		select {
		case <-ticker.C:
			var wg sync.WaitGroup
			wg.Add(3)
			go func() {
				a.genCounters(ctx)
				wg.Done()
			}()
			go func() {
				a.genGauges(ctx)
				wg.Done()
			}()
			go func() {
				a.genTimers(ctx)
				wg.Done()
			}()
			wg.Wait()
			log.Printf("flushed %d counters, %d gauges, %d timers for agent %d\n", len(a.counterNames), len(a.gaugeNames), len(a.timerNames), a.id)
		case <-ctx.Done():
			ticker.Stop()
			return
		}
	}

}

func (a *Agent) done(ctx context.Context) bool {
	select {
	case <-ctx.Done():
		return true
	default:
		return false
	}
}

func (a *Agent) genCounters(ctx context.Context) {
	for _, name := range a.counterNames {
		a.statsdClient.Count(name, rand.Intn(10))
		if a.done(ctx) {
			break
		}
	}
}

func (a *Agent) genGauges(ctx context.Context) {
	for _, name := range a.gaugeNames {
		a.statsdClient.Gauge(name, rand.Intn(500))
		if a.done(ctx) {
			break
		}
	}
}

func (a *Agent) genTimers(ctx context.Context) {
	for _, name := range a.timerNames {
		a.statsdClient.Timing(name, rand.Intn(1000))
		if a.done(ctx) {
			break
		}
	}
}

//flushOnce is to facilitate controlled testing
func (a *Agent) flushOnce() {
	a.statsdClient.Flush()
}

func genMetricsNames(metricType string, id, n int) []string {
	names := make([]string, n)
	for i := 0; i < n; i++ {
		names[i] = fmt.Sprintf("agent%d-%s%d", id, metricType, i)
	}
	return names
}

func parseTags(t string) (statsd.Option, error) {
	kvp := strings.Split(t, ",")
	var tags []string
	for _, pairs := range kvp {
		kv := strings.Split(pairs, ":")
		for _, tag := range kv {
			if tag == "" {
				return nil, errors.New("incomplete key:value pairs")
			}
			tags = append(tags, tag)
		}
	}
	if len(tags) < 2 || len(tags)%2 != 0 {
		return nil, errors.New("incomplete key:value pairs")
	}
	return statsd.Tags(tags...), nil
}

func parseTagFormat(client *statsd.Client, tf string) (*statsd.Client, error) {
	if tf == "datadog" {
		return client.Clone(statsd.TagsFormat(statsd.Datadog)), nil
	}
	if tf == "influx" {
		return client.Clone(statsd.TagsFormat(statsd.InfluxDB)), nil
	}
	return nil, errors.New("unrecognized tag format")
}

func done(ctx context.Context) bool {
	select {
	case <-ctx.Done():
		return true
	default:
		return false
	}
}


sighndlr_bsdsol.go
// Copyright © 2020 Circonus, Inc. <support@circonus.com>
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.
//

// +build freebsd openbsd solaris darwin

// Signal handling for FreeBSD, OpenBSD, Darwin, and Solaris
// systems that have SIGINFO

package main

import (
	"context"
	"fmt"
	"log"
	"os"
	"os/signal"
	"runtime"

	"github.com/alecthomas/units"
	"golang.org/x/sys/unix"
)

//SignalNotifySetup sets up the signals and their channel
func SignalNotifySetup(ch chan os.Signal) {
	signal.Notify(ch, os.Interrupt, unix.SIGTERM, unix.SIGHUP, unix.SIGPIPE, unix.SIGINFO)
}

//HandleSignals handles exiting the program based on different signals
func HandleSignals(cancel context.CancelFunc, ch chan os.Signal) {
	const stackTraceBufferSize = 1 * units.MiB

	//pre-allocate a buffer for stacktrace
	buf := make([]byte, stackTraceBufferSize)

	for {
		select {
		case sig := <-ch:
			log.Printf("signal %s received\n", sig.String())
			switch sig {
			case os.Interrupt, unix.SIGTERM:
				cancel()
				log.Println("waiting for final metric flushes... press CTRL+C to exit without flushing")
				break
			case unix.SIGPIPE, unix.SIGHUP:
				// Noop
			case unix.SIGINFO:
				stackLen := runtime.Stack(buf, true)
				fmt.Printf("=== received SIGINFO ===\n*** goroutine dump...\n%s\n*** end\n", buf[:stackLen])
			default:
				log.Printf("signal %s unsupported", sig.String())
			}
		}
	}
}


sighndlr_windows.go
// Copyright © 2020 Circonus, Inc. <support@circonus.com>
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.
//

// Signal handling for Windows
// doesn't have SIGINFO, attempt to use SIGTRAP instead...

package main

import (
	"context"
	"fmt"
	"log"
	"os"
	"os/signal"
	"runtime"
	"syscall"

	"github.com/alecthomas/units"
)

//SignalNotifySetup sets up the signals and their channel
func SignalNotifySetup(ch chan os.Signal) {
	signal.Notify(ch, os.Interrupt, syscall.SIGTERM, syscall.SIGHUP, syscall.SIGPIPE, syscall.SIGTRAP)
}

//HandleSignals handles exiting the program based on different signals
func HandleSignals(cancel context.CancelFunc, ch chan os.Signal) {
	const stackTraceBufferSize = 1 * units.MiB

	//pre-allocate a buffer for stacktrace
	buf := make([]byte, stackTraceBufferSize)

	for {
		select {
		case sig := <-ch:
			log.Printf("signal %s received\n", sig.String())
			switch sig {
			case os.Interrupt, syscall.SIGTERM:
				cancel()
				log.Println("waiting for final metric flushes... press CTRL+C to exit without flushing")
				break
			case syscall.SIGPIPE, syscall.SIGHUP:
				// Noop
			case syscall.SIGTRAP:
				stackLen := runtime.Stack(buf, true)
				fmt.Printf("=== received SIGINFO ===\n*** goroutine dump...\n%s\n*** end\n", buf[:stackLen])
			default:
				log.Printf("signal %s unsupported", sig.String())
			}
		}
	}
}


signlhndlr_linux.go
// Copyright © 2020 Circonus, Inc. <support@circonus.com>
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.
//

// Signal handling for Linux
// doesn't have SIGINFO, using SIGTRAP instead

package main

import (
	"context"
	"fmt"
	"log"
	"os"
	"os/signal"
	"runtime"

	"github.com/alecthomas/units"
	"golang.org/x/sys/unix"
)

//SignalNotifySetup sets up the signals and their channel
func SignalNotifySetup(ch chan os.Signal) {
	signal.Notify(ch, os.Interrupt, unix.SIGTERM, unix.SIGHUP, unix.SIGPIPE, unix.SIGTRAP)
}

//HandleSignals handles exiting the program based on different signals
func HandleSignals(cancel context.CancelFunc, ch chan os.Signal) {
	const stackTraceBufferSize = 1 * units.MiB

	//pre-allocate a buffer for stacktrace
	buf := make([]byte, stackTraceBufferSize)

	for {
		select {
		case sig := <-ch:
			log.Printf("signal %s received\n", sig.String())
			switch sig {
			case os.Interrupt, unix.SIGTERM:
				cancel()
				log.Println("waiting for final metric flushes... press CTRL+C to exit without flushing")
				break
			case unix.SIGPIPE, unix.SIGHUP:
				// Noop
			case unix.SIGTRAP:
				stackLen := runtime.Stack(buf, true)
				fmt.Printf("=== received SIGINFO ===\n*** goroutine dump...\n%s\n*** end\n", buf[:stackLen])
			default:
				log.Printf("signal %s unsupported", sig.String())
			}
		}
	}
}


config.go
// Copyright © 2020 Circonus, Inc. <support@circonus.com>
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.
//

package main

import (
	"log"
	"os"
	"path/filepath"
	"time"

	"github.com/namsral/flag"
)

type config struct {
	statsdHost    string
	prefix        string
	network       string
	tags          string
	tagFormat     string
	flushInterval time.Duration
	spawnDrift    int
	counters      int
	gauges        int
	timers        int
	agents        int
	version       bool
}

func genConfig() config {
	c := config{}

	defaultPrefix, err := os.Executable()
	if err != nil {
		log.Fatal(err)
	}
	flag.String(flag.DefaultConfigFlagname, "", "path to config file")
	flag.StringVar(&c.statsdHost, "statsd-host", "localhost:8125", "address of statsD host")
	flag.StringVar(&c.prefix, "prefix", filepath.Base(defaultPrefix), "prefix for metrics")
	flag.DurationVar(&c.flushInterval, "flush-interval", 10*time.Second, "how often to flush metrics")
	flag.IntVar(&c.spawnDrift, "spawn-drift", 10, "spread new agent generation by 0-n seconds")
	flag.StringVar(&c.network, "protocol", "udp", "network protocol to use, tcp or udp")
	flag.StringVar(&c.tagFormat, "tag-format", "", "format of the tags to send. accepted values \"datadog\" or \"influx\"")
	flag.StringVar(&c.tags, "tags", "", "list of K:V comma separated tags. Example: key1:tag1,key2:tag2")
	flag.IntVar(&c.counters, "counters", 50, "number of counters for each agent to hold")
	flag.IntVar(&c.gauges, "gauges", 30, "number of gauges for each agent to hold")
	flag.IntVar(&c.timers, "timers", 20, "number of timers for each agent to hold")
	flag.IntVar(&c.agents, "agents", 10, "max number of agents to run concurrently")
	flag.BoolVar(&c.version, "version", false, "show version information")
	flag.Parse()

	return c
}
